library/core/src/main/java/com/google/android/exoplayer2/audio/AudioCapabilities.java

```diff
@@ -158,4 +164,46 @@ public String toString() {
   private static boolean deviceMaySetExternalSurroundSoundGlobalSetting() {
     return Util.SDK_INT >= 17 && "Amazon".equals(Util.MANUFACTURER);
   }
+
+  @TargetApi(29)
+  private static int[] getDirectPlaybackSupportedEncodings() {
+      int[] possibleEncodings = new int[]{
+          AudioFormat.ENCODING_AAC_ELD,
+          AudioFormat.ENCODING_AAC_HE_V1,
+          AudioFormat.ENCODING_AAC_HE_V2,
+          AudioFormat.ENCODING_AAC_LC,
+          AudioFormat.ENCODING_AAC_XHE,
+          AudioFormat.ENCODING_AC3,
+          AudioFormat.ENCODING_AC4,
+          AudioFormat.ENCODING_DOLBY_TRUEHD,
+          AudioFormat.ENCODING_DTS,
+          AudioFormat.ENCODING_DTS_HD,
+          AudioFormat.ENCODING_E_AC3,
+          AudioFormat.ENCODING_E_AC3_JOC,
+          AudioFormat.ENCODING_IEC61937,
+          AudioFormat.ENCODING_MP3,
+          AudioFormat.ENCODING_PCM_8BIT,
+          AudioFormat.ENCODING_PCM_16BIT,
+          AudioFormat.ENCODING_PCM_FLOAT
+      };
+      List<Integer> supportedEncodingsList = new ArrayList<>();
+      for (int e : possibleEncodings) {
+          if (AudioTrack.isDirectPlaybackSupported(
+              new AudioFormat.Builder()
+                  .setChannelMask(AudioFormat.CHANNEL_OUT_STEREO)
```

Comment:
Do you know if the channel mask (and sample rate) are taken into account, or are they ignored?

Based on the marketing info MS12 [here](https://professional.dolby.com/product/home-entertainment/dolby-multistream-decoder-ms12/) I guess they aren't relevant because the MS12 decoder will convert the incoming audio (in any format) into one suitable for the current output capabilities, but I wanted to check in case you know otherwise.

Reply:
The audio framework requires channel mask and sample rate matching for direct outputs (it's searching for a compatible IOProfile and channel mask & sample rate are some of the parameters required to match). Checking only stereo / 48 kHz is not a complete solution, but is on par with the current AudioCapabilities.java implementation which is only checking for format support; this PS rests on the assumption that most direct outputs will support stereo / 48 kHz for all formats, and using stereo / 48 kHz is sufficient to extract the direct output format capabilities. A larger change in ExoPlayer checking for sample rate / channel mask matching is out of scope of this PS.

For MS12, all Dolby decoder (e.g. Dolby Digital Plus) implementations support all respective codec sample rates, and will decode to the current output capabilities.